### ISSUE TITLE
BLOCKS: Individual versioning? - Discussion

### ISSUE THREAD

**Issue by Whitevinyl** [BlokDust/BlokDust#290]
Just thinking about the backwards compatibility of blocks, currently with the few checks we're doing, we're looking up a specific outdated param and changing it. But what if we're still using the same param, but just the range or conversion has changed? It makes it much harder to look up what's changed.

If each block had a version number we could just reference that; if (this.Version < 2) {change stuff;}

We could just manually do it for now as each block version currently can be zero, and look at how to update it in future.
Another option would be to stamp each block with the app build number when created, which would be simpler to automate but not as nice to read/follow. This way may also leave us overly reliant on builds, so if you needed to do a block change and test it between builds that could be annoying.

Any other thoughts/ideas?

**Comment by lukephills** [comment]
Yeah I think we'll have to do this, seems like the simplest option. That switch statement could get messy after a while though but I'm not sure what else to suggest.

**Comment by edsilv** [comment]
We could possibly make the blocks independently versionable as npm modules. This could be really cool in that third parties would be able fork blocks and issue pull requests. An 'ecosystem' of blocks could emerge. The arguable downside of that is that it's not as easy to edit them as they're not part of one project any more. NPM has a pretty nice workflow though and my sense is that the project might  have more to gain by going down this path.

I think we should use releases of BD itself for backwards compatibility conversions. Each release of BD is a container of individual block versions.

We could write compat converters for each release. These could go in /app/compat or something, e.g. /app/compat/v1.5.ts These would each be responsible for converting between the previous and current release. We could then chain them together to convert from any previous release to the current one. A file like /app/compat/convert.ts would be part of the composition load pipeline, which checks the version number of the loaded comp, then dynamically builds an array of converters from that version to the current and executes them in sequence.

**Comment by Whitevinyl** [comment]
Yeah, making a separate backwards compatibility section/manager of BD is definitely a good idea.

Less sure about making the blocks npm, the third party benefits would be nice, but I know Luke & I are near constantly making small changes to blocks, taking them out of the project seems like it might be a big development hinderance.

The conversion chaining sounds great though. I think at this stage we should basically make sure we're laying simple foundations to build on this stuff, so how about we:

- add a manual individual version (1) to all blocks
- move all existing block patching to a backwards compatibility manager
- stamp composition blobs with the App build number both when created or overwritten

I think that will give us an easy way to manage it short-term, and the composition build number gives us a point to build future conversion chaining.

**Comment by edsilv** [comment]
Yeah, blocks over npm could be something we revisit later once they've "matured".

1. Manually versioning blocks sounds ok.

2. I think we'd definitely want to make the compatibility manager at least for release 2. I'm sure we'll need it!

3. Already done! https://github.com/BlokDust/BlokDust/blob/master/app/Serializer.ts#L48

**Comment by Whitevinyl** [comment]
Ha - was hoping you'd have that to say, re: point 3!
